### PR TITLE
[Validator] Feature #36362 add Isin validator constraint

### LINKED ISSUE #36362: ISIN code validator
There is a BIC constraint and validator, same about IBAN constraint and validator.

What do you think about adding an ISIN constraint and validator ?

Validation rule can be found here: https://en.wikipedia.org/wiki/International_Securities_Identification_Number

### PR DESCRIPTION
**Pull request by lmasforne**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #36362 
| License       | MIT
| Doc PR        | no

Add ISIN validator constraint 
see here for more details https://en.wikipedia.org/wiki/International_Securities_Identification_Number

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        $number = implode('', $characters);
+        $expectedCheckDigit = $this->getCheckDigit($number);
+
+        return $checkDigit === $expectedCheckDigit;
+    }
+
+    private function getCheckDigit($input)
+    {
+        // this method performs the luhn algorithm
+        // to obtain a check digit
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/Constraints/IsinValidator.php:92]: This description should be in the doc block of the method

File: CHANGELOG-5.0.md [hunk 2]
```diff
@@ -7,6 +7,8 @@ in 5.0 minor versions.
 To get the diff for a specific change, go to https://github.com/symfony/symfony/commit/XXX where XXX is the change hash
 To get the diff between two versions, go to https://github.com/symfony/symfony/compare/v5.0.0...v5.0.1
 
+ * feature add isin constraints in validator component (lmasforne)
+
```
Comment (YaFou) [inline CHANGELOG-5.0.md:11]: Your pull request is related to a new feature (branch `master`), so you have to write this change in [the changelog of the Validator Component (such as line 12)](https://github.com/symfony/symfony/blob/7f53ca55977a463882eded6d0f22b4836d1e3244/src/Symfony/Component/Validator/CHANGELOG.md).
```suggestion
```
This file provides patches on the minor release `5.0` (`5.0.1`, `5.0.2`, ...). When version `5.1` will be released, a new file called `CHANGELOG-5.1.md` will be created.

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        }
+
+        if (!is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))) {
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        $value = strtoupper($value);
+        $canonicalize = str_replace(' ', '', $value);
+
+        if (!(Isin::VALIDATION_LENGTH === \strlen($canonicalize))) {
```
Comment (YaFou) [inline src/Symfony/Component/Validator/Constraints/IsinValidator.php:46]: Why don't you use `!==`?
```suggestion
        if (Isin::VALIDATION_LENGTH !== \strlen($canonicalize)) {
```

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        return $checkDigit === $expectedCheckDigit;
+    }
+
+    /**
+     * This method performs the luhn algorithm
+     * to obtain a check digit.
+     */
+    private function getCheckDigit($input)
+    {
+        $input = (string) $input;
```
Comment (YaFou) [inline src/Symfony/Component/Validator/Constraints/IsinValidator.php:95]: For me, this line is unnecessary because it's a private function and the `$number` variable is always a sting (see [this](https://www.php.net/manual/en/function.implode.php)).

**Review by YaFou — CHANGES_REQUESTED** [review]
**Interesting pull request!** Just a few minor requests otherwise for me, this pull request is ready to be merged :rocket:

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php (hunk 4 above)
Comment (lmasforne) [reply src/Symfony/Component/Validator/Constraints/IsinValidator.php]: Yes it's right, i have fixed this

**Review by YaFou — APPROVED** [review]

File: CHANGELOG-5.0.md [hunk 5]
```diff
@@ -266,7 +266,7 @@ To get the diff between two versions, go to https://github.com/symfony/symfony/c
  * bug #34802 [Security] Check UserInterface::getPassword is not null before calling needsRehash (dbrekelmans)
  * bug #34788 [SecurityBundle] Properly escape regex in AddSessionDomainConstraintPass (fancyweb)
  * bug #34859 [SecurityBundle] Fix TokenStorage::reset not called in stateless firewall (jderusse)
- * bug #34827 [HttpFoundation] get currently session.gc_maxlifetime if ttl doesnt exists (rafaeltovar)
+ * bug #34827 [HttpFoundation] get currently session.gc_maxlifetime if ttl doesn't exists (rafaeltovar)
```
Comment (xabbuh) [inline CHANGELOG-5.0.md:269]: Please revert this change. The file is generated automatically based on the used commit messages.

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (!is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))) {
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        $value = strtoupper($value);
+        $canonicalize = str_replace(' ', '', $value);
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/IsinValidator.php:44]: I am not convinced that we should by default normalize the values to be normalized. In the worst scenario this will result in a loss of data when whitespaces are stripped here, but preserved in the original data meaning that any length constraint on the database level will raise issues (in the best case there will be an exception, but it gets worse if the database layer silently trims the data).

File: src/Symfony/Component/Validator/Resources/translations/validators.en.xlf [hunk 7]
```diff
@@ -382,6 +382,10 @@
                 <source>Each element of this collection should satisfy its own set of constraints.</source>
                 <target>Each element of this collection should satisfy its own set of constraints.</target>
             </trans-unit>
+            <trans-unit id="99">
+                <source>This is not a valid International Securities Identification Number (ISIN).</source>
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Resources/translations/validators.en.xlf:386]: "This value is not [...]" to be inline with other messages?

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php (hunk 6 above)
Comment (lmasforne) [reply src/Symfony/Component/Validator/Constraints/IsinValidator.php:44]: yes you're right, i remove this line, and it will lauch a validator violation ;)

File: CHANGELOG-5.0.md (hunk 5 above)
Comment (lmasforne) [reply CHANGELOG-5.0.md:269]: If i revert this `fabbot.io` is on error

Comment (OskarStark) [reply CHANGELOG-5.0.md:269]: It’s fine to revert, it’s a false positive 👍🏻

Comment (stof) [reply CHANGELOG-5.0.md:269]: @fabpot could you exclude these auto-generated files from the fabbot check ?

Comment (lmasforne) [reply CHANGELOG-5.0.md]: ok i have reverted the file thanks for your feedbacks

File: src/Symfony/Component/Validator/CHANGELOG.md [hunk 8]
```diff
@@ -9,6 +9,7 @@ CHANGELOG
  * allow to define a reusable set of constraints by extending the `Compound` constraint
  * added `Sequentially` constraint, to sequentially validate a set of constraints (any violation raised will prevent further validation of the nested constraints)
  * added the `divisibleBy` option to the `Count` constraint
+ * added `isin` constraints
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/CHANGELOG.md:12]:
```suggestion
 * added `Isin` constraint
```

**Comment by VincentLanglet** [comment]
Hi @lmasforne
Thanks for this PR ! Do you have some time to finish it ? :)

**Comment by lmasforne** [comment]
@VincentLanglet seems to be ok ;) ; hope this PR will be merged soon

**Comment by VincentLanglet** [comment]
I think you have to fix this first @lmasforne 
https://fabbot.io/report/symfony/symfony/36368/46116925704192e42fd76f0f86fad8b60e028f65

**Comment by VincentLanglet** [comment]
@lmasforne Your PR have some CI-failures

You need to fix https://fabbot.io/report/symfony/symfony/36368/46116925704192e42fd76f0f86fad8b60e028f65

Do you need help ?

**Comment by fabpot** [comment]
merged as #37565
